🌱 : Improve samples linting & fix samples lint issues

wazery opened the pull request:
✨ Summary

This PR enhances code quality and maintainability by addressing various linter issues and improving the CronJob controller Reconcile function code structure to reduce the cyclomatic complexity.

### What's included in the PR:

- Adds linting to documentation samples to ensure consistency and prevent future issues.
- Fixes lint issues across multiple tutorials and utilities, improving code readability and compliance.
- Refactors the CronJob controller’s Reconcile function for the samples to reduce cyclomatic complexity and enhance maintainability.
- Consolidates sample linting into a single workflow for better efficiency and organization.

### Motivation:

The goal of this PR is to enhance the overall linting coverage of sample code, ensuring that all included examples meet best practices. By addressing lint issues and integrating more samples, this change improves code consistency.

Fixes: #4445

wazery commented:
I hope my changes to the [cronjob_controller.go](https://github.com/kubernetes-sigs/kubebuilder/pull/4548/files#diff-3a8ecef9076f453232af2c24bc04badac070895640ce013bf163d1fbc02642dd) don't introduce any issues with the docs. I am trying to generate them and validate in the meantime.

camilamacedo86 commented on docs/book/src/multiversion-tutorial/testdata/project/internal/controller/cronjob_controller_test.go:
```diff
@@ -129,7 +129,7 @@ var _ = Describe("CronJob controller", func() {
 			Consistently(func(g Gomega) {
 				g.Expect(k8sClient.Get(ctx, cronjobLookupKey, createdCronjob)).To(Succeed())
-				g.Expect(createdCronjob.Status.Active).To(HaveLen(0))
+				g.Expect(createdCronjob.Status.Active).To(BeEmpty())
```
Hey, great work! 🎉

This looks really good, and you’ve already figured out what needs to be changed. 👏

Just a quick note:
All the samples under the docs (inside the book directory) are generated automatically when we run make generate. This command updates both the samples in testdata and the ones in docs because it calls make generate-docs.

So, instead of modifying the files directly in docs, the changes should be made in hack/docs first. After that, you’ll need to run:

sh
Copy
Edit
make generate-docs
to generate the final result correctly.

Could you take a look and check if you can apply the changes this way? It seems like you might be able to streamline some of the modifications as well.

Great job overall! 🚀 Looking forward to your update.

wazery replied:
Thanks a lot for the clarification 🙏 . Will reflect that on the PR accordingly.

camilamacedo86 replied:
If you run make generate-docs locally, you will see that all your changes here will be lost.
If we allow triggering the CI, you will see that it fails because we check to ensure that the changes are always properly generated.

camilamacedo86 commented on docs/book/src/simple-external-plugin-tutorial/testdata/sampleexternalplugin/v1/Makefile:
```diff
@@ -1,3 +1,5 @@
+# Image URL to use all building/pushing image targets
+IMG ?= controller:latest
```
That’s interesting! 🤔

I think the plugin could definitely have more targets—many of them would be valuable here, though not all from the default scaffold.

Just a quick note: this sample is actually an external plugin, meaning it will be packaged as a binary that generates scaffolds, similar to Kubebuilder. However, it is designed to work alongside Kubebuilder to extend functionality rather than replace its core scaffolding. More details here: [External Plugins](https://book.kubebuilder.io/plugins/extending/external-plugins).

Given that, I’d recommend reverting the changes in this PR.
If you’d like to improve the plugin by adding meaningful targets, that can definitely be done—but in a separate PR. We’ll also need to carefully evaluate which targets make sense for this context.

For example, we should not include targets like docker build for building a container image, because this project does not include a manager, Dockerfile, or containerized components, and it shouldn’t require them.

Hope this clarifies! Let me know your thoughts. 🚀

wazery replied:
Yes that truly makes perfect sense now to me. Will update the PR accordingly, thanks for the review.

wazery commented:
@camilamacedo86 updated the PR accordingly now. Truly appreciated the kind guidance 🙇 .

camilamacedo86 commented on hack/docs/internal/cronjob-tutorial/webhook_implementation.go:
```diff
@@ -205,7 +205,7 @@ const webhookTestCreateDefaultingReplaceFragment = `It("Should apply defaults wh
 
 			By("calling the Default method to apply defaults")
-			defaulter.Default(ctx, obj)
+			_ = defaulter.Default(ctx, obj)
```
Now, you need run `make generate-docs` to have the docs updated accordingly
If not that will fails in the ci.
I mean, if the result of the `make generaate-docs` is != of the docs then it will not pass.

wazery replied:
Just ran and amended the commit. thought this would be done automatically later by a bot or so 😅

camilamacedo86 replied:
That is because your branch is outdated.
Can you please rebase it with the master branch and re-run?

wazery replied:
Thanks @camilamacedo86. I just rebased it now and re-ran the generate-docs task.

camilamacedo86 commented on docs/book/src/cronjob-tutorial/testdata/project/internal/controller/cronjob_controller.go:
```diff
@@ -111,13 +112,12 @@ func (r *CronJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+	}
+	if cronJob == nil {
+		return ctrl.Result{}, nil // CronJob deleted
 	}
```
We cannot remove the previous content or the comments it is part of your docs/tutorials.
We just need to address the lint issues but keep all as before. 
What lint issue are you facing here?

camilamacedo86 commented on docs/book/src/cronjob-tutorial/testdata/project/internal/controller/cronjob_controller.go:
```diff
@@ -325,78 +251,15 @@ func (r *CronJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
-		}
-		return lastMissed, sched.Next(now), nil
-	}
-	// +kubebuilder:docs-gen:collapse=getNextSchedule
```
We will be able to make those changes here as well because this content is used un the tutorial
If the problem here is the complexity because the method is too long. Then, the best approach is to add on top of the fucn //nolint:complex <- The name of the linter to allow it to pass. 

Look the tutorial it will broke the docs.

wazery replied:
Addressed it, and used: `// nolint:gocyclo` for the Reconcile func. Thanks a lot for checking. Code is rebased on latest master.

camilamacedo86 reviewed (changes requested):
Hi @wazery 

Great work 🥇 
Just one thing here shows not accurate
We cannot change the `docs/book/src/cronjob-tutorial/testdata/project/internal/controller/cronjob_controller.go` at this away because it is used in the docs tutorial and those will broke it out.

I understand that the lint issue that you tried to solve is that the func has more than 30 lines, in this case we need add on top of the func `// nolint:<name of lint>` to ignore this case.

camilamacedo86 commented:
/ok-to-test
/approve
/lgtm

Great work 🥇 🚀 

Just I nit I will change the emoji because we use it to generate the release.
Since it is related to the docs I will add `:book:` but could also be 🌱

wazery commented:
I am not sure why the Testdata verification workflow fails, although when I run locally `make check-docs` it passes.

camilamacedo86 commented:
Hi @wazery 

I think it still not properly updated.
Can you please:
- Update your branch with the master
- Run `make install` && `make generate`
- Commit the changes / squash and push?


The validation just run the commands and check if has a git diff.
If the results are not == then it fails.

wazery commented:
> Hi @wazery
> 
> I think it still not properly updated. Can you please:
> 
> * Update your branch with the master
> * Run `make install` && `make generate`
> * Commit the changes / squash and push?
> 
> The validation just run the commands and check if has a git diff. If the results are not == then it fails.

I recreated the branch on latest master, reran the make install & generate commands and committed. Hope now it passes.

camilamacedo86 reviewed (approved):
/lgtm

camilamacedo86 commented:
Hi @wazery 

The lint has failures
And looking the diff ijn the testdata, it does not seems that your branch has the latest updates introduced in the master branch. You need like

git checkout master
git pull <remote name> master
git push

git checkout this-branch
git rebase master
make install
make generate

wazery commented:
@camilamacedo86 I fixed the issue, it was mainly that I introduced a run for goimports to reformat the file after generating it, but it seems this is an overkill. I removed that and rather fixed the format issue manually in the scripts so it will generate properly formatted code.

camilamacedo86 reviewed (approved):
/lgtm
